lib/src/ui/pages/authentication/login/login_state.dart

```diff
@@ -0,0 +1,38 @@
+import 'package:allthenews/src/domain/communication/all_the_news_exception.dart';
+
+class LoginState {
+  final bool isLoading;
+  final AllTheNewsException exception;
+  final String email;
+  final String emailError;
+  final String password;
+  final String passwordError;
+
+  bool get canSubmit => [emailError, passwordError].every((element) => element == null);
+
+  const LoginState({
+    this.email = "",
+    this.emailError,
+    this.password = "",
+    this.passwordError,
+    this.isLoading = false,
+    this.exception,
+  });
+
+  LoginState copyWith({
```

Comment:
trochę problematyczne, że część pól dopiero jak coś dostanie to updatuje, a exception i errory zawsze, mocno niespójne

Reply:
Chciałem, żeby domyślna wartość dla błędu to był `null` i żeby nullem można było też resetować błąd. Niestety nie znalazłem sposobu, żeby wartość przekazywana do funkcji jako parametr mogła być domyślnie wartością pola w klasie. Chciałem to zrobić tak jak działa `copy` w kotlinie, czyli w ten sposób :

![image](https://user-images.githubusercontent.com/24854180/101986632-3dd08e00-3c8f-11eb-9dd7-2fd8014a1be7.png)

ale wychodzi na to, że jedynym sensownym rozwiązaniem będzie przypisanie do errorów domyślnie pustych stringów (albo innej wartości, która będzie oznaczała brak błędu), bo nulle są w darcie zawsze domyślnie przekazywane i nie można na nich polegać jako domyślnych wartościach. Chyba, że masz jakiś inny pomysł, albo może przejście na `null safety`, które jest obecnie w becie rozwiąże ten problem?

Reply:
No tak po ludzku to raczej nie da rady tego zrobić, generalnie jest libka od tego gościa co zrobił providera, która ma takie kotlinowe copy: https://pub.dev/packages/freezed, ale generalnie znów jakiś code generator i szczerze mówiąc imo nie warto. Null safety raczej też nie rozwiąże tego problemu. Imo najprościej zrobić te pare metod, copyWithEmail, copyWithPassword, copyWithValidationErrors itd. które będą updatowały to co potrzebują updatować i może nie jest to idealne, no ale język chyba nic lepszego nie oferuje. Co o tym myślisz?

Reply:
bo zakładam, że tego samego podejścia używalibyśmy w innych projektach, więc dobrze sobie teraz ułatwić życie i ten sam schemat wszędzie już mieć ogarnięty :)

Reply:
https://github.com/google/quiver-dart/pull/672 widzę, że ten typ będzie wylatywał z tej libki, zastanawiam się czy rzeczywiście jakoś nie będzie można ograć tego lepiej jak przerzucimy się na null safety 🤔  dobra póki co zrobię każdą metodę osobną i taska na migrację do null safety, można w sumie już teraz przejść na darta betę i to ogarnąć

Reply:
Tak, można osobny task na null safety jak najbardziej. Co do optional to myślałem o tym, ale stwierdziłem, że byłoby to użycie niezgodne z przeznaczeniem bo jednak polega na tym że wartość jest lub jej nie ma i bardzo nieładne byłoby Some(null) :D Null safety za dużo nie da, bo jak obiekt będzie nullable to też będzie problem z rozróżnieniem, czy null został przekazany bo miał być czy po prostu jest defaultową wartością. Ciekawy ten patent z funkcją :D Ogólnie można stworzyć jakąś własną klasę, pokombinować z extensionami, żeby jak najmniej bolało ale to też w ramach osobnego taska, bo w sumie problem taki powszechny. Nie przeglądałem źródła freezed, kojarzyłem, że popularna, ale no dużo tych code generatorów to są rzeczy, które bierzesz bo wstyd by ci było je solo pisać xD